pkg/tracing/otlp_sdk.go

```diff
@@ -0,0 +1,118 @@
+package tracing
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"go.opentelemetry.io/contrib/exporters/autoexport"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.25.0"
+
+	"github.com/go-logr/logr"
+	errors "github.com/zgalor/weberr"
+)
+
+// ConfigureOpenTelemetryTracer configures the global OpenTelemetry trace
+// provider.
+//
+// The function uses the following environment variables for the tracer
+// configuration:
+//   - `OTEL_TRACES_EXPORTER`, either `otlp` to send traces to an OTLP endpoint or `console`.
+//   - `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL`, either `grpc` or `http`.
+//   - `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`, endpoint where to send the OTLP
+//     traces (e.g. `https://localhost:4318/v1/traces`).
+//
+// See
+// https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp
+// for the list of all supported variables.
+//
+// An error is returned if an environment value is set to an unhandled value.
+//
+// If no environment variable are set, a no-op tracer is setup.
+func ConfigureOpenTelemetryTracer(ctx context.Context, logger logr.Logger, serviceName string, serviceVersion string, cloudProvider string, resourceAttrs ...attribute.KeyValue) (func(context.Context) error, error) {
+	logger.Info("initializing OpenTelemetry tracer...")
+
+	exp, err := autoexport.NewSpanExporter(ctx, autoexport.WithFallbackSpanExporter(newNoopFactory))
+	if err != nil {
+		return nil, errors.Errorf("failed to create OTEL exporter: %s", err)
+	}
+
+	var isNoop bool
+	if _, isNoop = exp.(*noopSpanExporter); !isNoop || autoexport.IsNoneSpanExporter(exp) {
+		isNoop = true
+	}
+	logger.Info("initializing OpenTelemetry tracer:", "isNoop", isNoop)
+
+	opts := []resource.Option{resource.WithHost()}
+	if len(resourceAttrs) > 0 {
+		opts = append(opts, resource.WithAttributes(resourceAttrs...))
+	}
+	opts = append(opts, resource.WithAttributes(
+		semconv.ServiceNameKey.String(serviceName),
+		semconv.ServiceVersionKey.String(serviceVersion),
+		semconv.CloudProviderKey.String(cloudProvider),
```

Comment:
I wouldn't make the cloud provider a mandatory attribute, the caller can provide it in `resourceAttrs`.